File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/repository/InsertBuilder.java
package com.lasthopesoftware.bluewater.repository;

import java.util.ArrayList;

/**
 * Created by david on 12/14/15.
 */
public class InsertBuilder {
	private final StringBuilder sqlStringBuilder;
	private final ArrayList<String> columns = new ArrayList<>();

	public static InsertBuilder fromTable(String tableName) {
		return new InsertBuilder(tableName);
	}

	private InsertBuilder(String tableName) {
		sqlStringBuilder = new StringBuilder("INSERT INTO " + tableName + " (");
	}

	public InsertBuilder addColumn(String column) {
		columns.add(column);

		return this;
	}

	public String build() {
		for (String column : columns) {
			sqlStringBuilder.append(column);
			if (column != columns.get(columns.size() - 1))
				sqlStringBuilder.append(", ");
		}

		sqlStringBuilder.append(") VALUES (");

		for (String column : columns) {
			sqlStringBuilder.append(':').append(column);
			if (column != columns.get(columns.size() - 1))
				sqlStringBuilder.append(", ");
		}

		return sqlStringBuilder.append(')').toString();
	}
}

File: projectBlueWater/src/main/java/com/lasthopesoftware/bluewater/repository/InsertBuilder.kt
package com.lasthopesoftware.bluewater.repository

import java.util.*

/**
 * Created by david on 12/14/15.
 */
class InsertBuilder private constructor(private val tableName: String) {
	private val columns = ArrayList<String>()
	private var shouldReplace = false

	fun withReplacement(): InsertBuilder {
		shouldReplace = true
		return this
	}

	fun addColumn(column: String): InsertBuilder {
		columns.add(column)
		return this
	}

	fun build(): String {
		val sqlStringBuilder = StringBuilder("INSERT ");

		if (shouldReplace) sqlStringBuilder.append(" OR REPLACE ")

		sqlStringBuilder.append(" INTO $tableName (");

		for (column in columns) {
			sqlStringBuilder.append(column)
			if (column !== columns[columns.size - 1]) sqlStringBuilder.append(", ")
		}

		sqlStringBuilder.append(") VALUES (")
		for (column in columns) {
			sqlStringBuilder.append('@').append(column)
			if (column !== columns[columns.size - 1]) sqlStringBuilder.append(", ")
		}

		return sqlStringBuilder.append(')').toString()
	}

	companion object {
		@kotlin.jvm.JvmStatic
		fun fromTable(tableName: String): InsertBuilder {
			return InsertBuilder(tableName)
		}
	}
}
